File: common/worker/hasher/hasher.go
package hasher

import (
	"strings"
)

type HasherContainer struct {
	buffer map[string][][]string
}

// NewHasherContainer creates and initializes a new HasherContainer instance.
// It takes a map dictNodePosition, where each key is a string representing a node,
// and the value is an integer representing the number of replicas for that node.
// For each node, it allocates a slice of string slices (one for each replica) in the buffer.
// Returns a pointer to the newly created HasherContainer.
func NewHasherContainer(dictNodePosition map[string]int) *HasherContainer {
	buffer := make(map[string][][]string)

	for key, replicas := range dictNodePosition {
		replicaBuffer := make([][]string, replicas)
		for i := range replicas {
			replicaBuffer[i] = []string{}
		}
		buffer[key] = replicaBuffer
	}

	return &HasherContainer{
		buffer: buffer,
	}
}

// AddMessage adds a message associated with a given movieId to the buffer of each node type.
// The message is appended to the replica determined by movieId modulo the number of replicas for that node type.
// Parameters:
//   - movieId: an integer representing the ID of the movie, used to determine the replica index.
//   - message: the message string to be added to the buffer.
func (hc *HasherContainer) AddMessage(movieId int, message string) {
	for nodeType := range hc.buffer {
		replicaIndex := movieId % len(hc.buffer[nodeType])
		hc.buffer[nodeType][replicaIndex] = append(hc.buffer[nodeType][replicaIndex], message)
	}
}

// GetMessages retrieves and returns all buffered messages for the specified nodeType.
// The returned map has the message index as the key and the concatenated message string as the value.
// After retrieval, the messages for the given nodeType are cleared from the buffer.
// If there are no messages for the nodeType, it returns nil.
//
// Parameters:
//   - nodeType: The type of node whose messages are to be retrieved.
//
// Returns:
//   - map[int]string: A map where each key is the message index and the value is the concatenated message string.
func (hc *HasherContainer) GetMessages(nodeType string) map[int]string {
	if messages, exists := hc.buffer[nodeType]; exists {
		results := make(map[int]string)
		for i, msg := range messages {
			if len(msg) > 0 {
				results[i] = strings.Join(msg, "\n")
			}
		}
		hc.clearMessages(nodeType)
		return results
	}
	return nil
}

func (hc *HasherContainer) clearMessages(nodeType string) {
	if _, exists := hc.buffer[nodeType]; exists {
		for i := range hc.buffer[nodeType] {
			hc.buffer[nodeType][i] = []string{}
		}
	}
}

File: client-handler/common/hasher.py
from collections import defaultdict
import logging


class HasherContainer:
    def __init__(self, dict_node_type_positions: dict[str, int]):
        """
        Initializes the object with buffers for each node type.

        Args:
            dict_node_type_positions (dict[str, int]): 
                A dictionary mapping node types (as strings) to the number of positions (as integers) for each node type.

        Attributes:
            buffers (dict[str, list[list]]): 
                A dictionary where each key is a node type and each value is a list of empty lists, 
                with the number of lists corresponding to the specified positions for that node type.
        """
        self.buffers = {}
        for node_type, positions in dict_node_type_positions.items():
            self.buffers[node_type] = [[] for _ in range(positions)]

    def append_to_node(self, movie_id: int, data: str):
        """
        Appends the given data to the appropriate buffer shard for each node type based on the movie ID.

        Args:
            movie_id (int): The unique identifier for the movie, used to determine the shard index.
            data (str): The data to append to the selected buffer shard.

        Details:
            For each node type in self.buffers, calculates the shard index by taking the modulo of movie_id
            with the number of shards for that node type. Appends the data to the corresponding shard.
        """
        for node_type in self.buffers:
            shard_idx = movie_id % len(self.buffers[node_type])
            self.buffers[node_type][shard_idx].append(data)

    def get_buffers(self) -> dict[str, dict[int, str]]:
        """
        Aggregates and retrieves the current buffer contents for each node type and shard.

        Returns:
            dict[str, dict[int, str]]: A dictionary where each key is a node type (str),
            and each value is another dictionary mapping shard indices (int) to the
            concatenated string contents of their respective buffers. After retrieval,
            the buffers for each shard are cleared.
        """
        results = defaultdict(dict)

        for node_type, lists in self.buffers.items():
            for i, shard in enumerate(lists):
                if not shard:
                    continue
                results[node_type][i] = "".join(shard)
                self.buffers[node_type][i] = []

        return dict(results)
